### PR TITLE
tikv-ctl: Add multi-threaded recover-mvcc

### PR DESCRIPTION
**Pull request by MyonKeminta**
## What have you changed? (mandatory)

This PR added a subcommand `recover-mvcc-all` to tikv-ctl. 

It does the same thing as `recover-mvcc` subcommand, which deletes those keys that doesn't compose an MVCC key. For example, if a `PUT` record in WRITE_CF requires a corresponding record in DEFAULT_CF. If there's not such a key in DEFAULT_CF, the key in  WRITE_CF will be deleted.

The difference between `recover-mvcc-all` and `recover-mvcc` is that, first, `recover-mvcc-all` works on a whole offline RocksDB while the old `recover-mvcc` runs on a region; second, `recover-mvcc-all` supports multi-threading; third, it supports the read-only mode, in which it only prints out the invalid keys but doesn't really fix it.

## What are the type of the changes? (mandatory)

- New feature (non-breaking change which adds functionality)

## How has this PR been tested? (mandatory)

- By unit tests and manual tests

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)

To be done.

## Does this PR affect tidb-ansible update? (mandatory)

No

### CODE REVIEW

File: src/server/debug.rs [hunk 1]
```diff
@@ -360,6 +371,97 @@ impl Debugger {
         Ok(())
     }
 
+    pub fn recover_all(&self, threads: usize, read_only: bool) -> Result<()> {
```
Comment (zhangjinpeng1987) [inline src/server/debug.rs]: Please add a test for this function. Load some corrupted data into local RocksDB and recover it.

File: src/server/debug.rs [hunk 2]
```diff
@@ -360,6 +371,97 @@ impl Debugger {
...
+                .name(format!("mvcc-recover-thread-{}", thread_index))
+                .spawn(move || {
+                    println!(
+                        "thread {}: started on range [\"{}\", \"{}\")",
+                        thread_index,
+                        escape(&start_key),
+                        escape(&end_key)
+                    );
+
+                    let mut mvcc_checker =
```
Comment (Connor1996) [inline src/server/debug.rs]: this part is verbose with `recover_region`, you can extract them into `recover_range`

File: src/server/debug.rs [hunk 3]
```diff
@@ -1163,6 +1287,86 @@ fn set_region_tombstone(db: &DB, store_id: u64, region: Region, wb: &WriteBatch)
...
+    }
+
+    if keys.len() > 20000 {
+        let len = keys.len();
+        keys = keys.into_iter().step_by(len / 10000).collect();
+    }
+
+    keys.sort();
+
+    Ok(sample(keys, parts - 1))
```
Comment (Hijiao) [inline src/server/debug.rs]: could `parts` be 0 here?

Comment (MyonKeminta) [reply src/server/debug.rs]: Ouch. I'll filter it out.

File: src/server/debug.rs (hunk 1 above)
Comment (MyonKeminta) [reply src/server/debug.rs]: The recovering logic is already tested in `test_mvcc_checker` in debug.rs

**Comment by MyonKeminta** [comment]
@disksing @Connor1996 PTAL

File: src/server/debug.rs [hunk 4]
```diff
@@ -322,42 +325,87 @@ impl Debugger {
...
-            mvcc_checker.default_fix_count,
-            mvcc_checker.lock_fix_count,
-            mvcc_checker.write_fix_count
-        );
+        let mut handles = Vec::new();
 
-        Ok(())
+        for thread_index in 0..range_borders.len() - 1 {
+            let db = Arc::clone(db);
+            let start_key = range_borders[thread_index].clone();
```
Comment (Connor1996) [inline src/server/debug.rs]: you can use `slice::chunks`

File: src/server/debug.rs [hunk 5]
```diff
@@ -1165,6 +1276,86 @@ fn set_region_tombstone(db: &DB, store_id: u64, region: Region, wb: &WriteBatch)
...
+    }
+
+    if samples >= data.len() {
+        return data;
+    }
+
+    let len = data.len();
+    data.into_iter()
+        .enumerate()
+        .filter(|(i, _)| i + 1 != len && i * (samples + 1) / len != (i + 1) * (samples + 1) / len)
```
Comment (disksing) [inline src/server/debug.rs]: It takes me some time to understand it. 😸

**Comment by disksing** [comment]
LGTM.

File: src/server/debug.rs (hunk 5 above)
Comment (MyonKeminta) [reply src/server/debug.rs]: This is used to find all dividing points and I have no idea how to simplify the calculation..

File: src/server/debug.rs (hunk 4 above)
Comment (MyonKeminta) [reply src/server/debug.rs]: It seems it's not useful here.

File: src/server/debug.rs [hunk 6]
```diff
@@ -322,42 +325,87 @@ impl Debugger {
...
-            mvcc_checker.write_fix_count
-        );
+        let mut handles = Vec::new();
 
-        Ok(())
+        for thread_index in 0..range_borders.len() - 1 {
+            let db = Arc::clone(db);
+            let start_key = range_borders[thread_index].clone();
+            let end_key = range_borders[thread_index + 1].clone();
+            if start_key == end_key {
```
Comment (Connor1996) [inline src/server/debug.rs]: what if split_keys is empty

File: src/server/debug.rs [hunk 7]
```diff
@@ -1165,6 +1276,86 @@ fn set_region_tombstone(db: &DB, store_id: u64, region: Region, wb: &WriteBatch)
...
+    fake_region.set_start_key(vec![]);
+    fake_region.set_end_key(vec![]);
+    fake_region.mut_peers().push(Peer::default());
+    let get_cf_size =
+        |cf: &str| raftstore_util::get_region_approximate_size_cf(db, cf, &fake_region);
+
+    let default_cf_size = box_try!(get_cf_size(CF_DEFAULT));
+    let write_cf_size = box_try!(get_cf_size(CF_WRITE));
+
+    let middle_by_cf = if default_cf_size >= write_cf_size {
```
Comment (Connor1996) [inline src/server/debug.rs]:
```suggestion
     let cf = if default_cf_size >= write_cf_size {
```

File: src/server/debug.rs [hunk 8]
```diff
@@ -1165,6 +1276,86 @@ fn set_region_tombstone(db: &DB, store_id: u64, region: Region, wb: &WriteBatch)
...
+        CF_DEFAULT
+    } else {
+        CF_WRITE
+    };
+
+    divide_db_cf(db, parts, middle_by_cf)
```
Comment (Connor1996) [inline src/server/debug.rs]:
```suggestion
     divide_db_cf(db, parts, cf)
```

**Comment by disksing** [comment]
LGTM

**Comment by MyonKeminta** [comment]
@disksing  PTAL again

**Pull request merged**